<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>trim the reads with special apdater and barocde | Hexo</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="read结构R1 5’→3’: genome_sequenceR2 5’→3’: special_adapter + barcode1 + linker1 + barcode2 + linker2 + barcode3 + genome_sequence Rawdata文件中的read1和read2序列通常去除了P5和P7接头。而在本次我处理的read文件中，read2的5’端加上了一段特定的额外">
<meta property="og:type" content="article">
<meta property="og:title" content="trim the reads with special apdater and barocde">
<meta property="og:url" content="http://example.com/2024/03/03/test/index.html">
<meta property="og:site_name" content="Hexo">
<meta property="og:description" content="read结构R1 5’→3’: genome_sequenceR2 5’→3’: special_adapter + barcode1 + linker1 + barcode2 + linker2 + barcode3 + genome_sequence Rawdata文件中的read1和read2序列通常去除了P5和P7接头。而在本次我处理的read文件中，read2的5’端加上了一段特定的额外">
<meta property="og:locale" content="en_US">
<meta property="article:published_time" content="2024-03-03T07:33:41.000Z">
<meta property="article:modified_time" content="2024-03-14T06:36:32.680Z">
<meta property="article:author" content="John Doe">
<meta property="article:tag" content="single-cell">
<meta property="article:tag" content="pretreatment">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="Hexo" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
  
<meta name="generator" content="Hexo 6.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">Hexo</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"><span class="fa fa-bars"></span></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
        
          <a class="nav-icon" href="/atom.xml" title="RSS Feed"><span class="fa fa-rss"></span></a>
        
        <a class="nav-icon nav-search-btn" title="Search"><span class="fa fa-search"></span></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="Search"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://example.com"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main"><article id="post-test" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2024/03/03/test/" class="article-date">
  <time class="dt-published" datetime="2024-03-03T07:33:41.000Z" itemprop="datePublished">2024-03-03</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/single-cell/">single-cell</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 class="p-name article-title" itemprop="headline name">
      trim the reads with special apdater and barocde
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="read结构"><a href="#read结构" class="headerlink" title="read结构"></a>read结构</h1><p>R1 5’→3’: genome_sequence<br>R2 5’→3’: special_adapter + barcode1 + linker1 + barcode2 + linker2 + barcode3 + genome_sequence</p>
<p>Rawdata文件中的read1和read2序列通常去除了P5和P7接头。<br>而在本次我处理的read文件中，read2的5’端加上了一段特定的额外接头序列(seq1)，而在这一段接头序列之后又连接了单细胞特异的barcode结构： barcode1 + linker1 + barcode2 + linker2 + barcode3。</p>
<h2 id="处理思路"><a href="#处理思路" class="headerlink" title="处理思路"></a>处理思路</h2><ol>
<li>过滤read pair, 只提取read2 5’端具有 seq1 的 read pair。这一部分的工作可以由cutadapt完成。</li>
<li>过滤 第1步中处理好的 reads 文件，筛选其中 5’端 具有特定barcode结构的read pair，并获取read ID 和 barcode 之间的对应关系。这一部分的工作应该可以用UMItools完成，但我这次打算先自己写脚本处理。</li>
<li>过滤低质量，短长度的reads。这一部分的工作可以用fastp完成。</li>
</ol>
<h2 id="预期得到的文件"><a href="#预期得到的文件" class="headerlink" title="预期得到的文件"></a>预期得到的文件</h2><ol>
<li>经过trim和高质量的 有效read的R1和R2文件</li>
<li>Reads ID与 barcode 对应的文件</li>
</ol>
<h2 id="实际操作"><a href="#实际操作" class="headerlink" title="实际操作"></a>实际操作</h2><h3 id="cutadapt去除特定的接头序列"><a href="#cutadapt去除特定的接头序列" class="headerlink" title="cutadapt去除特定的接头序列"></a>cutadapt去除特定的接头序列</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ sample_ID=XXX</span><br><span class="line">$ adapter=XXXXXXXXXXX</span><br><span class="line"></span><br><span class="line">$ cutadapt -G <span class="variable">$&#123;adapter&#125;</span> -j 10 -e 0.1 -O 5 -m 50 \ </span><br><span class="line">--pair-filter=both  --discard-untrimmed -o <span class="variable">$&#123;sample_ID&#125;</span>_cutadapt_R1.fq.gz \ </span><br><span class="line">-p <span class="variable">$&#123;sample_ID&#125;</span>_cutadapt_R2.fq.gz <span class="variable">$&#123;sample_ID&#125;</span>_R1.fq.gz  <span class="variable">$&#123;sample_ID&#125;</span>_R2.fq.gz</span><br></pre></td></tr></table></figure>
<p>cutadapt命令解释</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">-G #指定R2中5&#x27;端的接头序列， 相似的参数还有-g,-a,-A</span><br><span class="line">-j #运行该程序的核数</span><br><span class="line">-e #接头序列与reads序列匹配过程中允许的最大错配率(error rate)</span><br><span class="line">-O #接头序列与reads序列匹配过程中允许的最小重叠长度(overlap)</span><br><span class="line">-m #trim接头之后，如果reads的长度小于该值，则将该条reads丢弃(min-length)</span><br><span class="line">--pair-filter=both #如果read1或read2中的某条序列被丢弃了，那么它对应的read2或read1也同时被丢弃</span><br><span class="line">--discard-untrimmed #将没有被trim的reads丢弃，在本命令中的意思即：若reads2 5&#x27;端无配对得上$&#123;adapter&#125;得序列，则将该reads2丢弃，又因为 --pair-filter=both， 对应的read1也被丢弃</span><br><span class="line">-o, -p #输出裁剪后的对应的R1和R2文件</span><br><span class="line">最后两个就是输入文件</span><br></pre></td></tr></table></figure>


<p>cutdapt的使用可以参考<a target="_blank" rel="noopener" href="https://cutadapt.readthedocs.io/en/stable/guide.html">cutadapt官方文档</a>或者<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/4ee2f4d2292f">cutdapt使用详解</a></p>
<h3 id="鉴定具有barcode结构的reads-pair"><a href="#鉴定具有barcode结构的reads-pair" class="headerlink" title="鉴定具有barcode结构的reads pair"></a>鉴定具有barcode结构的reads pair</h3><ol>
<li><p>定义READ类用以存储fastq数据，因为每一个READ ID都对应两条READ，所以这个类中存储的序列信息都是配对存在的。具体的脚本可以参考[identify_barcode.py]</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">READ</span>:</span><br><span class="line">    <span class="comment">#define the READ object to store the information of paired reads</span></span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__init__</span>(<span class="params"> self , ID:<span class="built_in">str</span></span>) -&gt; <span class="literal">None</span>:</span><br><span class="line">        self.<span class="built_in">id</span> = ID</span><br><span class="line">        self.seq = [ <span class="string">&#x27;&#x27;</span> , <span class="string">&#x27;&#x27;</span>]</span><br><span class="line">        self.anno = [ <span class="string">&#x27;&#x27;</span> , <span class="string">&#x27;&#x27;</span>]</span><br><span class="line">        self.quality = [ <span class="string">&#x27;&#x27;</span> , <span class="string">&#x27;&#x27;</span> ]</span><br><span class="line">        self.barcode = <span class="string">&#x27;&#x27;</span></span><br><span class="line">        self.valid = <span class="literal">False</span></span><br><span class="line">        <span class="keyword">pass</span></span><br></pre></td></tr></table></figure></li>
<li><p>从fastq文件中读取对应ID的read1与read2的序列与其他信息。gzip.open相对比较慢，用了据说比较快的pyfastx读取fq.gz文件。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">def</span> <span class="title function_">read_paired_fq_file</span>(<span class="params"> r1_file:<span class="built_in">str</span> , r2_file:<span class="built_in">str</span> , limit:<span class="built_in">int</span> = math.inf</span>):</span><br><span class="line">    reads_dic = &#123;&#125;</span><br><span class="line">    file_cnt = <span class="number">0</span>     <span class="comment">#mark which file is loading, read1 or read2 </span></span><br><span class="line">    <span class="keyword">for</span>  tmp_file <span class="keyword">in</span> [ r1_file , r2_file ]:</span><br><span class="line">        read_cnt = <span class="number">0</span></span><br><span class="line">        file_cnt += <span class="number">1</span></span><br><span class="line">        <span class="keyword">for</span> read <span class="keyword">in</span> pyfastx.Fastq( tmp_file ):</span><br><span class="line">            tmp_id = read.name[:-<span class="number">2</span>]</span><br><span class="line">            <span class="keyword">if</span> reads_dic.get( tmp_id):</span><br><span class="line">                reads_dic[ tmp_id ].add_read( seq =read.seq, anno= <span class="string">&#x27;+&#x27;</span> ,  quality = read.qual , kind = file_cnt )</span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                reads_dic[ tmp_id ] = READ( tmp_id )</span><br><span class="line">                reads_dic[ tmp_id].add_read( seq =read.seq, anno= <span class="string">&#x27;+&#x27;</span> ,  quality = read.qual , kind = file_cnt )</span><br><span class="line">            read_cnt += <span class="number">1</span></span><br><span class="line">            <span class="keyword">if</span> read_cnt &gt; limit:</span><br><span class="line">                <span class="keyword">break</span></span><br><span class="line">    <span class="keyword">return</span> reads_dic</span><br></pre></td></tr></table></figure>
</li>
<li><p>定义判断序列间是否匹配的函数，并根据barcode结构，编写拆分序列的脚本，并一一匹配，判断reads是否有效与对应的barcode编码</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#允许错配的函数</span></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">mispair_match</span>(<span class="params"> ref_s:<span class="built_in">str</span> , query_s:<span class="built_in">str</span> , limit:<span class="built_in">int</span> = <span class="number">1</span></span>) -&gt; <span class="built_in">bool</span>:</span><br><span class="line">    mis = <span class="number">0</span></span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>( <span class="built_in">len</span>(ref_s) ):</span><br><span class="line">        <span class="keyword">if</span> ref_s[i] == query_s[i]:</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            mis += <span class="number">1</span></span><br><span class="line">    <span class="keyword">return</span> mis &lt;= limit</span><br></pre></td></tr></table></figure></li>
<li><p>输出所有符合要求的reads分别到R1和R2文件中。上述所有脚本都可在[]找到</p>
</li>
</ol>
<p>最后的命令行调用</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Python identify_barcode.py -1 <span class="variable">$&#123;sample_ID&#125;</span>_cutadapt_R1.fq.gz  -2 <span class="variable">$&#123;sample_ID&#125;</span>_cutadapt_R2.fq.gz \</span><br><span class="line">-b 0127_barcode_pool  \</span><br><span class="line">-l TGGT,GAGA  \</span><br><span class="line">-o1 01.data/<span class="variable">$&#123;sample_ID&#125;</span>_filtered_R1   -o2 01.data/<span class="variable">$&#123;sample_ID&#125;</span>_filtered_R2 </span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">-1,-2 #输入的R1和R2的fq文件</span><br><span class="line">-b #barcode库文件，不同位置的barcode库用##分割</span><br><span class="line">-l #linker序列，中间用,隔开</span><br><span class="line">-o1,-o2 #输出的R1和R2文件前缀</span><br></pre></td></tr></table></figure>

<h2 id="Fastp过滤reads"><a href="#Fastp过滤reads" class="headerlink" title="Fastp过滤reads"></a>Fastp过滤reads</h2><p>主要就是去除一些裁剪后比较短的序列和低质量的reads。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">fastp \</span><br><span class="line">--in1 <span class="variable">$&#123;sampleID&#125;</span>_R1.fq.gz \</span><br><span class="line">--in2 <span class="variable">$&#123;sampleID&#125;</span>_R2.fq.gz \</span><br><span class="line">--out1 <span class="variable">$&#123;sampleID&#125;</span>_clean_R1.fq.gz \</span><br><span class="line">--out2 <span class="variable">$&#123;sampleID&#125;</span>_clean_R2.fq.gz \</span><br><span class="line">--json <span class="variable">$&#123;sampleID&#125;</span>.json \</span><br><span class="line">--html <span class="variable">$&#123;sampleID&#125;</span>.html \</span><br><span class="line">--trim_poly_g --poly_g_min_len 10 \</span><br><span class="line">--trim_poly_x --poly_x_min_len 10 \</span><br><span class="line">--cut_front --cut_tail --cut_window_size 4 \</span><br><span class="line">--qualified_quality_phred 30 \</span><br><span class="line">--low_complexity_filter \</span><br><span class="line">--complexity_threshold 30 \</span><br><span class="line">--length_required 30 \</span><br><span class="line">--thread 4</span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">--in1, --in2, --out1, --out2 #输入与输出文件</span><br><span class="line">--json, --html  #过滤报告输出文件</span><br><span class="line">--trim_poly_g, --trim_poly_x    # trim read两端的poly G/X</span><br><span class="line">--poly_g_min_len, --poly_x_min_len   #定义poly G/X的最短长度</span><br><span class="line"></span><br><span class="line">--cut_front, --cut_tail     #从5&#x27;端 和 3&#x27;端裁剪低质量或低复杂度的窗口</span><br><span class="line">--cut_window_size    # 定义窗口大小</span><br><span class="line">--qualified_quality_phred  #定义窗口内的低质量碱基的阈值</span><br><span class="line">--low_complexity_filter   #对于低复杂度的窗口进行过滤</span><br><span class="line">--complexity_threshold   #定义低复杂度的阈值</span><br><span class="line"></span><br><span class="line">--length_required    #定义裁剪后序列的最短长度，若是小于该值，则舍弃该序列</span><br><span class="line"></span><br><span class="line">--thread   #定义运行该程序的核数量</span><br><span class="line"></span><br></pre></td></tr></table></figure>
      
    </div>
    <footer class="article-footer">
      <a data-url="http://example.com/2024/03/03/test/" data-id="cltqz7rjr0002b4vmfqbifllv" data-title="trim the reads with special apdater and barocde" class="article-share-link"><span class="fa fa-share">Share</span></a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/pretreatment/" rel="tag">pretreatment</a></li><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/single-cell/" rel="tag">single-cell</a></li></ul>

    </footer>
  </div>
  
    
<nav id="article-nav">
  
  
    <a href="/2024/03/03/hello-world/" id="article-nav-older" class="article-nav-link-wrap">
      <strong class="article-nav-caption">Older</strong>
      <div class="article-nav-title">Flying_doggy&#39;s home page</div>
    </a>
  
</nav>

  
</article>


</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Categories</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/single-cell/">single-cell</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Tags</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/pretreatment/" rel="tag">pretreatment</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/single-cell/" rel="tag">single-cell</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Tag Cloud</h3>
    <div class="widget tagcloud">
      <a href="/tags/pretreatment/" style="font-size: 10px;">pretreatment</a> <a href="/tags/single-cell/" style="font-size: 10px;">single-cell</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Archives</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2024/03/">March 2024</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Recent Posts</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2024/03/03/test/">trim the reads with special apdater and barocde</a>
          </li>
        
          <li>
            <a href="/2024/03/03/hello-world/">Flying_doggy&#39;s home page</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2024 John Doe<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.6.4.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>