<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.3.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Muse","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="Introduction这篇文章主要用于记录转录组数据分析的流程与要点，尽可能全面准确地描述整个分析过程与每一步的意义。转录组的数据目前通常是通过提取组织mRNA并逆转录为cDNA，再通过二代测序获得的短片段（150bp）的序列。 由于转录过程可能涉及到 可变剪切 ， 转录本融合 等一系列转录后修饰事件，通常需要注释文件（gtf,gff）辅助基因组建立索引，可以使用STAR , Hisat2 等软">
<meta property="og:type" content="article">
<meta property="og:title" content="transcriptome_analysis">
<meta property="og:url" content="http://example.com/2024/05/01/transcriptome-analysis/index.html">
<meta property="og:site_name" content="Hexo">
<meta property="og:description" content="Introduction这篇文章主要用于记录转录组数据分析的流程与要点，尽可能全面准确地描述整个分析过程与每一步的意义。转录组的数据目前通常是通过提取组织mRNA并逆转录为cDNA，再通过二代测序获得的短片段（150bp）的序列。 由于转录过程可能涉及到 可变剪切 ， 转录本融合 等一系列转录后修饰事件，通常需要注释文件（gtf,gff）辅助基因组建立索引，可以使用STAR , Hisat2 等软">
<meta property="og:locale" content="en_US">
<meta property="article:published_time" content="2024-05-01T13:06:53.000Z">
<meta property="article:modified_time" content="2024-05-01T14:48:42.782Z">
<meta property="article:author" content="John Doe">
<meta property="article:tag" content="transcriptome analysis">
<meta property="article:tag" content="STAR">
<meta property="article:tag" content="RSEM">
<meta property="article:tag" content="DESeq">
<meta property="article:tag" content="omiverse">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/2024/05/01/transcriptome-analysis/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>transcriptome_analysis | Hexo</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Hexo</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="http://example.com/2024/05/01/transcriptome-analysis/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.gif">
      <meta itemprop="name" content="John Doe">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Hexo">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          transcriptome_analysis
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2024-05-01 21:06:53 / Modified: 22:48:42" itemprop="dateCreated datePublished" datetime="2024-05-01T21:06:53+08:00">2024-05-01</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/transcriptome/" itemprop="url" rel="index"><span itemprop="name">transcriptome</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <h1 id="Introduction"><a href="#Introduction" class="headerlink" title="Introduction"></a>Introduction</h1><p>这篇文章主要用于记录转录组数据分析的流程与要点，尽可能全面准确地描述整个分析过程与每一步的意义。<br>转录组的数据目前通常是通过提取组织mRNA并逆转录为cDNA，再通过二代测序获得的短片段（150bp）的序列。</p>
<p>由于转录过程可能涉及到 <strong>可变剪切</strong> ， <strong>转录本融合</strong> 等一系列转录后修饰事件，通常需要注释文件（gtf,gff）辅助基因组建立索引，可以使用<a target="_blank" rel="noopener" href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf">STAR</a> , Hisat2 等软件对转录组索引进行构建；如没有注释文件的话，也可以用Hisat2直接构建转录组索引，但可能比对的效果会差一些。</p>
<p>将原始数据中的每一条reads比对到对应的转录本后，我们就可以根据每条转录本上的reads数量对该转录本的表达水平进行定量。常用的软件有<a target="_blank" rel="noopener" href="https://github.com/deweylab/RSEM">RSEM</a>, StringTie。</p>
<blockquote>
<p>而不同批次间的差异与转录本的长度会影响到reads比对到该转录本上的数量（长转录本在测序过程中被捕获到的数量往往会更多些，测序量更大的实验组，捕获的reads数量也会更多），因此常根据这两项对转录本的定量进行标准化，常见的为FPKM，RPKM, TPM。</p>
</blockquote>
<p>当我们获得了每个样本的转录表达谱后，我们可以对这个高维矩阵进行 <strong>主成分分析</strong> ，鉴定不同实验组之间的差异性。同时可以根据自己的实验设计，对不同组别间某些基因的表达情况进行差异分析，共表达分析等。在本次实验中，我们主要进行了 <strong>差异表达基因分析</strong> 。可以用热图，曼哈顿图等多种方式展示差异表达情况。</p>
<p>筛选出来的基因可以根据需求进行功能富集分析，motif分析等来找到导致其差异表达的因素。</p>
<h1 id="01-Install"><a href="#01-Install" class="headerlink" title="01. Install"></a>01. Install</h1><h2 id="1-1-转录组比对软件STAR"><a href="#1-1-转录组比对软件STAR" class="headerlink" title="1.1 转录组比对软件STAR"></a>1.1 转录组比对软件STAR</h2><p>可以直接克隆<a target="_blank" rel="noopener" href="https://github.com/alexdobin/STAR/">STAR</a>在Github上的工程到本地并开始使用，也可以下载安装包后编译使用。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#直接克隆后就可以使用</span></span><br><span class="line">git <span class="built_in">clone</span> https://github.com/alexdobin/STAR.git</span><br></pre></td></tr></table></figure>

<h2 id="1-2-安装数据质控软件FastQC"><a href="#1-2-安装数据质控软件FastQC" class="headerlink" title="1.2 安装数据质控软件FastQC"></a>1.2 安装数据质控软件FastQC</h2><p>直接在<a target="_blank" rel="noopener" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">官网</a>上下载压缩包，解压后就可以使用FastQC。是我个人过滤高通量数据常用的软件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.12.1.zip</span><br><span class="line">unzip fastqc_v0.12.1.zip</span><br></pre></td></tr></table></figure>

<h2 id="1-3-安装转录本定量软件RSEM"><a href="#1-3-安装转录本定量软件RSEM" class="headerlink" title="1.3 安装转录本定量软件RSEM"></a>1.3 安装转录本定量软件RSEM</h2><p>RSEM还需要一些基础的安装环境，不知道为什么我的Unbuntu上居然没有，大家用自己的服务器可以直接make，不许要前面的apt install</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Unbuntu中确实相应的库，一般不用跑这两步</span></span><br><span class="line">sudo apt install build-essential   </span><br><span class="line">sudo apt install zlib1g-dev</span><br><span class="line"></span><br><span class="line"><span class="comment"># 常规的安装步骤</span></span><br><span class="line">git <span class="built_in">clone</span> https://github.com/deweylab/RSEM.git</span><br><span class="line">make &amp; make install DESTDIR=/home/gujianhui/bio_tools/RSEM  prefix=/RSEM</span><br></pre></td></tr></table></figure>

<h1 id="02-Preparation"><a href="#02-Preparation" class="headerlink" title="02.Preparation"></a>02.Preparation</h1><h2 id="2-1-建立转录组索引"><a href="#2-1-建立转录组索引" class="headerlink" title="2.1 建立转录组索引"></a>2.1 建立转录组索引</h2><p>可以从数据库中下载相应的基因组数据与基因组注释文件（转录组比对比较推荐gtf文件），也可以直接下载STAR官方提供的一些转录组索引</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 若数据库只提供gff文件，可以使用gffread将gff文件转化为gtf文件格式</span></span><br><span class="line">gffread -T gff_file -o gtf_file</span><br><span class="line"></span><br><span class="line"><span class="comment"># STAR建立比对索引(使用gtf建立索引)</span></span><br><span class="line"><span class="variable">$&#123;path_to_STAR&#125;</span>/STAR \</span><br><span class="line">--runThreadN 20 \</span><br><span class="line">--runMode genomeGenerate \</span><br><span class="line">--genomeDir ./   \</span><br><span class="line">--genomeFastaFiles <span class="variable">$&#123;path_to_genome&#125;</span>/TAIR10.fa  \</span><br><span class="line">--sjdbGTFfile <span class="variable">$&#123;path_to_gtf&#125;</span>/TAIR10.gtf \</span><br><span class="line">--sjdbOverhang 149</span><br><span class="line"></span><br><span class="line"><span class="comment"># or STAR建立比对索引(使用gff建立索引)</span></span><br><span class="line"><span class="variable">$&#123;path_to_STAR&#125;</span>/STAR \</span><br><span class="line">--runThreadN 20 \</span><br><span class="line">--runMode genomeGenerate \</span><br><span class="line">--genomeDir ./   \</span><br><span class="line">--genomeFastaFiles <span class="variable">$&#123;path_to_genome&#125;</span>/TAIR10.fa  \</span><br><span class="line">--sjdbGTFtagExonParentTranscript Parent  \</span><br><span class="line">--sjdbGTFfile <span class="variable">$&#123;path_to_gff&#125;</span>/TAIR10.gff \</span><br><span class="line">--sjdbOverhang 149</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>具体每个参数的意义可以参考 <a target="_blank" rel="noopener" href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf">STAR手册</a>或者我之前的文章。<code>--clip3pNbases</code> 参数可以跳过测序质量较差的开头几个碱基，提高比对效率，但会在bam文件中造成软跳过，无法用后续的RSEM进行定量</p>
<p>除此之外，使用RSEM进行转录本的定量也需要建立对应的索引。<a target="_blank" rel="noopener" href="https://open.bioqueue.org/home/knowledge/showKnowledge/sig/rsem-prepare-reference">rsem-prepare-reference</a></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">rsem-prepare-reference --gtf <span class="variable">$&#123;path_to_gtf&#125;</span>/TAIR10.gtf \</span><br><span class="line">-p 10  \</span><br><span class="line"><span class="variable">$&#123;path_to_genome&#125;</span>/TAIR10.fa  \</span><br><span class="line">rsem_idx/rsem</span><br></pre></td></tr></table></figure>
<p>在使用过程中发现，rsem-prepare-reference也可以同时建立star，bowtie2等其他常见比对软件的比对索引。建议直接 <code>rsem-prepare-reference -h</code> 查看。</p>
<h2 id="2-2-测序数据过滤"><a href="#2-2-测序数据过滤" class="headerlink" title="2.2 测序数据过滤"></a>2.2 测序数据过滤</h2><p>使用FastQC，fastp对原始测序数据进行评估与过滤，去除其中测序质量较差或存在接头污染的reads pair。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="variable">$&#123;path_to_fastqc&#125;</span>/fastqc -t 20  \</span><br><span class="line">-o <span class="variable">$&#123;output_dir&#125;</span>  \</span><br><span class="line">*/*.fq.gz</span><br></pre></td></tr></table></figure>
<p>根据质控html报告对测序结果进行裁减过滤等操作，目前的质量一般都还比较OK</p>
<h1 id="03-Map-Quantification"><a href="#03-Map-Quantification" class="headerlink" title="03. Map &amp; Quantification"></a>03. Map &amp; Quantification</h1><p>将处理好的reads文件比对到建立好转录组索引，获取比对结果的toTranscriptome.bam文件（需要设置<code>--quantMode TranscriptomeSAM</code> ）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="variable">$&#123;path_to_star&#125;</span>/STAR  \</span><br><span class="line">--runMode alignReads \</span><br><span class="line">--runThreadN 12 \</span><br><span class="line">--readFilesIn <span class="variable">$&#123;path_to_data&#125;</span>/*.fq.gz \</span><br><span class="line">--genomeDir <span class="variable">$&#123;STAR_index&#125;</span>  \</span><br><span class="line">--outSAMtype BAM SortedByCoordinate  \</span><br><span class="line">--twopassMode Basic \</span><br><span class="line">--readFilesCommand zcat  \</span><br><span class="line">--quantMode TranscriptomeSAM GeneCounts \</span><br><span class="line">--outFileNamePrefix  <span class="variable">$&#123;prefix&#125;</span>  \</span><br><span class="line">--outTmpDir  <span class="variable">$&#123;output_dir&#125;</span>  \</span><br></pre></td></tr></table></figure>
<p>对于所有reads的比对统计结果存储于 ${prefix}Log.final.out 文件中，包括了比对效率，Unique比对率等等。</p>
<p>若对STAR软件的结果文件比较熟悉的话，可以直接根据 ReadsPerGene.out.tab 文件进行分析，手动写脚本对转录本进行定量，可以参考<a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1936443">STAR GeneCounts结果定量</a>。也可以用现有的轮子RSEM对比对结果进行定量</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">rsem-calculate-expression --paired-end  \</span><br><span class="line">--no-bam-output --alignments -p 15 -q 02 \</span><br><span class="line"><span class="variable">$&#123;path_toTranscriptome_bam&#125;</span> \</span><br><span class="line"><span class="variable">$&#123;path_to_rsem_index&#125;</span> \</span><br><span class="line"><span class="variable">$&#123;output_dir&#125;</span></span><br></pre></td></tr></table></figure>
<p>通常，结果文件中的 genes.results就统计了每个基因&#x2F;转录本上的比对reads数量</p>
<h1 id="04-downstream-analysis"><a href="#04-downstream-analysis" class="headerlink" title="04.downstream_analysis"></a>04.downstream_analysis</h1><p>通常进行的转录组分析为差异表达（Differential Expression）分析，我们需要将之前得到的定量数据导入到分析软件&#x2F;环境中。在本次分析中，我主要在R环境中，使用DESeq对转录组结果进行分析。<br>首先，将所有结果文件整合为各实验组的转录表达矩阵。列名为每个实验组的标签，一行对应了一个基因在各个样本中的表达情况</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">rsem-generate-data-matrix *.genes.results &gt; merged.genes_result</span><br></pre></td></tr></table></figure>

<p>以下为简单的R分析脚本</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>DESeq2<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>ggplot2<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>ggrepel<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">exp_df <span class="operator">&lt;-</span> read.table<span class="punctuation">(</span><span class="string">&#x27;merged.genes_result&#x27;</span> <span class="punctuation">,</span> </span><br><span class="line">                     header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> </span><br><span class="line">                     row.names <span class="operator">=</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#调整每个单元格中的数据为整数</span></span><br><span class="line">exp_df <span class="operator">&lt;-</span><span class="built_in">round</span><span class="punctuation">(</span>exp_df<span class="punctuation">,</span>digits <span class="operator">=</span> <span class="number">0</span><span class="punctuation">)</span></span><br><span class="line">exp_mat <span class="operator">&lt;-</span> as.matrix<span class="punctuation">(</span> exp_df<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#对实验组数据进行分组</span></span><br><span class="line">groups <span class="operator">&lt;-</span> factor<span class="punctuation">(</span> <span class="built_in">c</span><span class="punctuation">(</span> <span class="built_in">rep</span><span class="punctuation">(</span><span class="string">&quot;H&quot;</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">)</span> <span class="punctuation">,</span></span><br><span class="line">                     <span class="built_in">rep</span><span class="punctuation">(</span><span class="string">&quot;S0&quot;</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">)</span> <span class="punctuation">,</span> <span class="built_in">rep</span><span class="punctuation">(</span><span class="string">&quot;S1&quot;</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">)</span> <span class="punctuation">,</span></span><br><span class="line">                     <span class="built_in">rep</span><span class="punctuation">(</span><span class="string">&quot;T0&quot;</span> <span class="punctuation">,</span><span class="number">3</span><span class="punctuation">)</span> <span class="punctuation">,</span><span class="built_in">rep</span><span class="punctuation">(</span><span class="string">&quot;T1&quot;</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">)</span> <span class="punctuation">)</span> <span class="punctuation">)</span></span><br><span class="line">col_group <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span> row.names <span class="operator">=</span> colnames<span class="punctuation">(</span>exp_df<span class="punctuation">)</span> <span class="punctuation">,</span> </span><br><span class="line">                         groups<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 生成DESeq对象，并将转录组结果与分组要求输入</span></span><br><span class="line">de_result <span class="operator">&lt;-</span> DESeqDataSetFromMatrix<span class="punctuation">(</span>countData<span class="operator">=</span>exp_mat<span class="punctuation">,</span></span><br><span class="line">                              colData<span class="operator">=</span>col_group<span class="punctuation">,</span></span><br><span class="line">                              design<span class="operator">=</span><span class="operator">~</span>groups<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 归一化数据等</span></span><br><span class="line">de_result <span class="operator">&lt;-</span> DESeq<span class="punctuation">(</span>de_result<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 对转录表达谱进行主成分分析，观察不同实验组之间的转录谱差异是否显著</span></span><br><span class="line">plotPCA<span class="punctuation">(</span> rlog<span class="punctuation">(</span>de_result<span class="punctuation">)</span> <span class="punctuation">,</span> intgroup<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;groups&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 比较指定两组之间的差异表达基因</span></span><br><span class="line">S0_S1_res <span class="operator">&lt;-</span> results<span class="punctuation">(</span> de_result <span class="punctuation">,</span></span><br><span class="line">                      contrast <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;groups&quot;</span><span class="punctuation">,</span><span class="string">&#x27;S0&#x27;</span><span class="punctuation">,</span><span class="string">&quot;S1&quot;</span><span class="punctuation">)</span> <span class="punctuation">,</span></span><br><span class="line">                      alpha<span class="operator">=</span><span class="number">0.05</span><span class="punctuation">)</span></span><br><span class="line">summary<span class="punctuation">(</span> TO_H_res <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 标记在分析过程中存在显著差异（上调，下调等）的基因</span></span><br><span class="line">S0_S1_res<span class="operator">$</span>sig <span class="operator">&lt;-</span> <span class="string">&#x27;no_sig&#x27;</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span> i <span class="keyword">in</span> <span class="number">1</span><span class="operator">:</span><span class="built_in">length</span><span class="punctuation">(</span>row.names<span class="punctuation">(</span>S0_S1_res<span class="punctuation">)</span> <span class="punctuation">)</span> <span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  <span class="keyword">if</span> <span class="punctuation">(</span> <span class="built_in">is.na</span><span class="punctuation">(</span>S0_S1_res<span class="operator">$</span>padj<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">)</span><span class="operator">==</span><span class="literal">FALSE</span> <span class="operator">&amp;</span> S0_S1_res<span class="operator">$</span>padj<span class="punctuation">[</span>i<span class="punctuation">]</span> <span class="operator">&lt;</span> <span class="number">0.05</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    <span class="keyword">if</span> <span class="punctuation">(</span> S0_S1_res<span class="operator">$</span>log2FoldChange<span class="punctuation">[</span>i<span class="punctuation">]</span> <span class="operator">&gt;</span> <span class="number">1</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">      S0_S1_res<span class="operator">$</span>sig<span class="punctuation">[</span>i<span class="punctuation">]</span> <span class="operator">=</span> <span class="string">&#x27;up&#x27;</span></span><br><span class="line">    <span class="punctuation">&#125;</span> <span class="keyword">else</span> <span class="keyword">if</span><span class="punctuation">(</span> S0_S1_res<span class="operator">$</span>log2FoldChange<span class="punctuation">[</span>i<span class="punctuation">]</span> <span class="operator">&lt;</span> <span class="operator">-</span><span class="number">1</span> <span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">      S0_S1_res<span class="operator">$</span>sig<span class="punctuation">[</span>i<span class="punctuation">]</span> <span class="operator">=</span> <span class="string">&#x27;down&#x27;</span></span><br><span class="line">    <span class="punctuation">&#125;</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 根据注释结果与差异分析结果绘制火山图，展示两组实验之间发生表达变化的基因</span></span><br><span class="line">ggplot<span class="punctuation">(</span>data<span class="operator">=</span>S0_S1_res<span class="punctuation">,</span> aes<span class="punctuation">(</span>x<span class="operator">=</span>log2FoldChange<span class="punctuation">,</span> y <span class="operator">=</span><span class="operator">-</span>log10<span class="punctuation">(</span>padj<span class="punctuation">)</span><span class="punctuation">,</span>color <span class="operator">=</span>sig<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  geom_point<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  ylim<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="number">8</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">  scale_color_manual<span class="punctuation">(</span>values <span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;no_sig&#x27;</span><span class="operator">=</span><span class="string">&quot;black&quot;</span><span class="punctuation">,</span></span><br><span class="line">                               <span class="string">&#x27;up&#x27;</span> <span class="operator">=</span> <span class="string">&quot;red&quot;</span><span class="punctuation">,</span></span><br><span class="line">                               <span class="string">&#x27;down&#x27;</span> <span class="operator">=</span> <span class="string">&quot;blue&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">  geom_hline<span class="punctuation">(</span>yintercept <span class="operator">=</span> log10<span class="punctuation">(</span><span class="number">0.05</span><span class="punctuation">)</span><span class="punctuation">,</span>lty<span class="operator">=</span><span class="number">4</span><span class="punctuation">,</span>lwd<span class="operator">=</span><span class="number">0.6</span><span class="punctuation">,</span>alpha<span class="operator">=</span><span class="number">0.8</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">  geom_vline<span class="punctuation">(</span>xintercept <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span>lty<span class="operator">=</span><span class="number">4</span><span class="punctuation">,</span>lwd<span class="operator">=</span><span class="number">0.6</span><span class="punctuation">,</span>alpha<span class="operator">=</span><span class="number">0.8</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">  theme_bw<span class="punctuation">(</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">  theme<span class="punctuation">(</span>panel.border <span class="operator">=</span> element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">        panel.grid.major <span class="operator">=</span> element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">        panel.grid.minor <span class="operator">=</span> element_blank<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">        axis.line <span class="operator">=</span> element_line<span class="punctuation">(</span>colour <span class="operator">=</span> <span class="string">&quot;black&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">  labs<span class="punctuation">(</span>title<span class="operator">=</span><span class="string">&quot;Volcanoplot_biotrainee S0 vs S1&quot;</span><span class="punctuation">,</span> x<span class="operator">=</span><span class="string">&quot;log2 (fold change)&quot;</span><span class="punctuation">,</span>y<span class="operator">=</span><span class="string">&quot;-log10 (padj)&quot;</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">  </span><br><span class="line">  theme<span class="punctuation">(</span>plot.title <span class="operator">=</span> element_text<span class="punctuation">(</span>hjust <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">  geom_text_repel<span class="punctuation">(</span>data<span class="operator">=</span>subset<span class="punctuation">(</span>S0_S1_res<span class="punctuation">,</span> <span class="operator">-</span>log10<span class="punctuation">(</span>padj<span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="number">6</span><span class="punctuation">)</span><span class="punctuation">,</span> aes<span class="punctuation">(</span>label<span class="operator">=</span>Gene<span class="punctuation">)</span><span class="punctuation">,</span>col<span class="operator">=</span><span class="string">&quot;black&quot;</span><span class="punctuation">,</span>alpha <span class="operator">=</span><span class="number">0.8</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>

    </div>

    
    
    

      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/transcriptome-analysis/" rel="tag"># transcriptome analysis</a>
              <a href="/tags/STAR/" rel="tag"># STAR</a>
              <a href="/tags/RSEM/" rel="tag"># RSEM</a>
              <a href="/tags/DESeq/" rel="tag"># DESeq</a>
              <a href="/tags/omiverse/" rel="tag"># omiverse</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2024/03/14/genome_assembly/" rel="prev" title="genome assembly">
      <i class="fa fa-chevron-left"></i> genome assembly
    </a></div>
      <div class="post-nav-item">
    <a href="/2024/09/26/ChIP-without-control/" rel="next" title="ChIP_without_control">
      ChIP_without_control <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#Introduction"><span class="nav-number">1.</span> <span class="nav-text">Introduction</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#01-Install"><span class="nav-number">2.</span> <span class="nav-text">01. Install</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#1-1-%E8%BD%AC%E5%BD%95%E7%BB%84%E6%AF%94%E5%AF%B9%E8%BD%AF%E4%BB%B6STAR"><span class="nav-number">2.1.</span> <span class="nav-text">1.1 转录组比对软件STAR</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#1-2-%E5%AE%89%E8%A3%85%E6%95%B0%E6%8D%AE%E8%B4%A8%E6%8E%A7%E8%BD%AF%E4%BB%B6FastQC"><span class="nav-number">2.2.</span> <span class="nav-text">1.2 安装数据质控软件FastQC</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#1-3-%E5%AE%89%E8%A3%85%E8%BD%AC%E5%BD%95%E6%9C%AC%E5%AE%9A%E9%87%8F%E8%BD%AF%E4%BB%B6RSEM"><span class="nav-number">2.3.</span> <span class="nav-text">1.3 安装转录本定量软件RSEM</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#02-Preparation"><span class="nav-number">3.</span> <span class="nav-text">02.Preparation</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#2-1-%E5%BB%BA%E7%AB%8B%E8%BD%AC%E5%BD%95%E7%BB%84%E7%B4%A2%E5%BC%95"><span class="nav-number">3.1.</span> <span class="nav-text">2.1 建立转录组索引</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-2-%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE%E8%BF%87%E6%BB%A4"><span class="nav-number">3.2.</span> <span class="nav-text">2.2 测序数据过滤</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#03-Map-Quantification"><span class="nav-number">4.</span> <span class="nav-text">03. Map &amp; Quantification</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#04-downstream-analysis"><span class="nav-number">5.</span> <span class="nav-text">04.downstream_analysis</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
  <p class="site-author-name" itemprop="name">John Doe</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">5</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
        <span class="site-state-item-count">3</span>
        <span class="site-state-item-name">categories</span>
      </div>
      <div class="site-state-item site-state-tags">
        <span class="site-state-item-count">10</span>
        <span class="site-state-item-name">tags</span>
      </div>
  </nav>
</div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2024</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">John Doe</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://muse.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Muse</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/muse.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

</body>
</html>
