<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.3.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Muse","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="Introduction这份报告主要用以记录在用长片段测序结果组装基因组的方案，流程以及遇到的一些困难与分析结果。 长片段测序的HIFI数据，常用hifiasm这一软件进行初步组装，完成Contig水平的基因组。 再补充HIC数据后，可以juicer+3D-DNA, yahs等软件对Contig再次进行组装，成为scaffold水平的基因组。在对软件拼接结果的手动调整后，可能可以得到染色体水平的基">
<meta property="og:type" content="website">
<meta property="og:title" content="genome assembly">
<meta property="og:url" content="http://example.com/categories/index.html">
<meta property="og:site_name" content="Hexo">
<meta property="og:description" content="Introduction这份报告主要用以记录在用长片段测序结果组装基因组的方案，流程以及遇到的一些困难与分析结果。 长片段测序的HIFI数据，常用hifiasm这一软件进行初步组装，完成Contig水平的基因组。 再补充HIC数据后，可以juicer+3D-DNA, yahs等软件对Contig再次进行组装，成为scaffold水平的基因组。在对软件拼接结果的手动调整后，可能可以得到染色体水平的基">
<meta property="og:locale" content="en_US">
<meta property="article:published_time" content="2024-03-14T06:27:59.000Z">
<meta property="article:modified_time" content="2024-03-14T08:35:28.333Z">
<meta property="article:author" content="John Doe">
<meta property="article:tag" content="genome assembly">
<meta property="article:tag" content="hifiasm">
<meta property="article:tag" content="juicer">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/categories/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : false,
    lang   : 'en'
  };
</script>

  <title>genome assembly | Hexo
</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Hexo</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
  
  

          <div class="content page posts-expand">
            

    
    
    
    <div class="post-block" lang="en">
      <header class="post-header">

<h1 class="post-title" itemprop="name headline">genome assembly
</h1>

<div class="post-meta">
  

</div>

</header>

      
      
      
      <div class="post-body">
          <h1 id="Introduction"><a href="#Introduction" class="headerlink" title="Introduction"></a>Introduction</h1><p>这份报告主要用以记录在用长片段测序结果组装基因组的方案，流程以及遇到的一些困难与分析结果。</p>
<p>长片段测序的HIFI数据，常用<a target="_blank" rel="noopener" href="https://github.com/chhylp123/hifiasm">hifiasm</a>这一软件进行初步组装，完成Contig水平的基因组。 再补充HIC数据后，可以juicer+3D-DNA, yahs等软件对Contig再次进行组装，成为scaffold水平的基因组。在对软件拼接结果的手动调整后，可能可以得到染色体水平的基因组文件。</p>
<p>tips: contig和scaffold的概念主要是在使用 <strong>二代测序</strong> 获得短片段组装基因组的时候提出的，这两者在HIC挂载后的长度差异较大。<br>但是用了 <strong>三代测序</strong> 的长片段后，这两者的长度差异小了许多,这两个概念相对模糊了一点</p>
<h1 id="1-Install-software"><a href="#1-Install-software" class="headerlink" title="1. Install software"></a>1. Install software</h1><p>在实际操作过程中，主要采用了hifiasm+juicer+3D-DNA的组装策略</p>
<h2 id="1-1-安装hifiasm"><a href="#1-1-安装hifiasm" class="headerlink" title="1.1 安装hifiasm"></a>1.1 安装hifiasm</h2><h3 id="1-1-1-conda安装"><a href="#1-1-1-conda安装" class="headerlink" title="1.1.1 conda安装"></a>1.1.1 conda安装</h3><p><code>conda install -c bioconda hifiasm </code></p>
<h3 id="1-1-2-github安装"><a href="#1-1-2-github安装" class="headerlink" title="1.1.2 github安装"></a>1.1.2 github安装</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ git <span class="built_in">clone</span> https://github.com/chhylp123/hifiasm</span><br><span class="line">$ <span class="built_in">cd</span> hifiasm &amp;&amp; make </span><br></pre></td></tr></table></figure>


<h2 id="1-2-安装AGB-（将gfa转为fa文件-可选）"><a href="#1-2-安装AGB-（将gfa转为fa文件-可选）" class="headerlink" title="1.2 安装AGB （将gfa转为fa文件,可选）"></a>1.2 安装AGB （将gfa转为fa文件,可选）</h2><p>hifiasm输出的文件为gfa组装文件，要将其转化为fasta文件便于后续分析，可以使用AGB。也可以自己写脚本去转一下格式。 笔者这边偷懒就用AGB了<br><code>conda create -c almiheenko -c bioconda -n AGB agb</code></p>
<h2 id="1-3-juicer安装"><a href="#1-3-juicer安装" class="headerlink" title="1.3 juicer安装"></a>1.3 juicer安装</h2><p>juicer相比于其他的HIC数据比对软件运行起来需要准备的文件更多一些，但一个命令行就可以把原始数据过滤，比对等一系列过程全部完成。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">$ git <span class="built_in">clone</span> https://github.com/aidenlab/juicer.git</span><br><span class="line">$ <span class="built_in">cd</span> juicer</span><br><span class="line">$ <span class="built_in">ln</span> -s juicer/CPU scripts</span><br><span class="line">$ <span class="built_in">cd</span> scripts/common</span><br></pre></td></tr></table></figure>

<p>以上步骤都和网络上的教程差不多，但部分教程会推荐安装juicer_tools 1.9版本(适配于Juicer1.6)，这可能会因为版本(git clone 下载的版本为juicer 2.0)不匹配，无法正常输出 <strong>inter.txt</strong> 文件，故需要安装juicer_tools 2.0及以上</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ wget -c https://github.com/aidenlab/Juicebox/releases/download/v2.20.00/juicer_tools.2.20.00.jar</span><br><span class="line">$ <span class="built_in">ln</span> -s juicer_tools.2.20.00.jar juicer_tools.jar</span><br></pre></td></tr></table></figure>
<p>参考<a target="_blank" rel="noopener" href="https://phantom-aria.github.io/2023/07/11/a.html">Hi-C染色体挂载（1）——juicer2处理Hi-C数据</a></p>
<blockquote>
<p>Github官方文档说明 <a target="_blank" rel="noopener" href="https://github.com/aidenlab/juicer">juicer</a><br>The beta release for Juicer version 1.6 can be accessed via the Github Release. The main repository on Github is now focused on the Juicer 2.0 release and is under active development. For general questions, please use the Google Group.</p>
</blockquote>
<h2 id="1-4-3D-DNA安装"><a href="#1-4-3D-DNA安装" class="headerlink" title="1.4 3D-DNA安装"></a>1.4 3D-DNA安装</h2><p>直接git clone就行 <a target="_blank" rel="noopener" href="https://github.com/aidenlab/3d-dna">3D-DNA</a><br><code>$ git clone https://github.com/theaidenlab/3d-dna.git</code></p>
<h1 id="2-基因组组装"><a href="#2-基因组组装" class="headerlink" title="2. 基因组组装"></a>2. 基因组组装</h1><p>所需文件： </p>
<ol>
<li>HIFI测序文件(fa.gz)</li>
<li>HIC测序文件（_1&#x2F;2.fq.gz）</li>
</ol>
<p>bash变量命名说明：</p>
<ul>
<li>${sp}   物种名称</li>
<li>${hic_sample}  hic测序文件前缀编号</li>
<li>${path_to_hic} hic文件所在路径</li>
<li>${path_to_hifi} hifi文件所在路径</li>
<li>${hifi_sample}   hifi测序文件前缀编号</li>
</ul>
<h2 id="2-1-hifiasm拼接"><a href="#2-1-hifiasm拼接" class="headerlink" title="2.1 hifiasm拼接"></a>2.1 hifiasm拼接</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ conda activate hifiasm</span><br><span class="line">$ hifiasm  --primary -o <span class="variable">$&#123;sp&#125;</span>_hifiasm \</span><br><span class="line">-t 20 \</span><br><span class="line">--h1 <span class="variable">$&#123;path_to_hic&#125;</span>/<span class="variable">$&#123;sp&#125;</span>_HIC_R1.fq.gz  \</span><br><span class="line">--h2 <span class="variable">$&#123;path_to_hic&#125;</span>/<span class="variable">$&#123;sp&#125;</span>_HIC_R2.fq.gz  \ </span><br><span class="line"><span class="variable">$&#123;path_to_hifi&#125;</span>/<span class="variable">$&#123;hifi_sample&#125;</span>.fa.gz &gt; assembly_log</span><br></pre></td></tr></table></figure>
<p>具体可以参考<a target="_blank" rel="noopener" href="https://hifiasm.readthedocs.io/en/latest/index.html">hifiasm文档</a></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">--primary  输出主要的组装结果</span><br><span class="line">-o 输出文件前缀</span><br><span class="line">-t threads数量，运行该命令的线程数</span><br><span class="line">--h1 HIC文件的read1文件</span><br><span class="line">--h2 HIC文件的read2文件</span><br></pre></td></tr></table></figure>
<p>最后得到的许多文件中，主要使用了p_ctg.gfa文件</p>
<ul>
<li>可选分析<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#统计每个contig的长度，进而计算N50, L50</span></span><br><span class="line">grep <span class="string">&#x27;^S&#x27;</span> <span class="variable">$&#123;sp&#125;</span>_hifiasm.hic.p_ctg.gfa |awk <span class="string">&#x27;&#123;print $2 , length($3)&#125;&#x27;</span></span><br></pre></td></tr></table></figure></li>
</ul>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#将gfa文件转化为fasta文件</span></span><br><span class="line">$ agb.py --graph <span class="variable">$&#123;sp&#125;</span>_hifiasm.p_ctg.gfa -a hifiasm</span><br><span class="line">$ <span class="built_in">mv</span> agb_output/<span class="variable">$&#123;sp&#125;</span>_hifiasm.p_ctg.fasta   <span class="variable">$&#123;sp&#125;</span>.hifi.fasta</span><br></pre></td></tr></table></figure>


<h2 id="2-2-juicer比对"><a href="#2-2-juicer比对" class="headerlink" title="2.2 juicer比对"></a>2.2 juicer比对</h2><p>在使用juicer对HIC数据进行比对前，需要准备一系列的文件：①酶切位点位置文件; ②基因组lens文件 ; ③基因组bwa索引 ; ④包含了HIC测序文件的fastq文件夹</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#生成基因组酶切位点位置文件</span></span><br><span class="line">Python <span class="variable">$&#123;path_to_juicer&#125;</span>/misc/generate_site_positions.py  DpnII  <span class="variable">$&#123;sp&#125;</span>  <span class="variable">$&#123;sp&#125;</span>.hifi.fasta</span><br><span class="line"></span><br><span class="line"><span class="comment">#生成基因组lens文件 , $&#123;sp&#125;_DpnII为上一步的结果文件，也可以用samtools faidx生成lens文件</span></span><br><span class="line">awk <span class="string">&#x27;BEGIN&#123;OFS=&quot;\t&quot;&#125;&#123;print $1, $NF&#125;&#x27;</span> <span class="variable">$&#123;sp&#125;</span>_DpnII.txt &gt; <span class="variable">$&#123;sp&#125;</span>.lens</span><br><span class="line"></span><br><span class="line"><span class="comment">#建立基因组bwa索引</span></span><br><span class="line">bwa index <span class="variable">$&#123;sp&#125;</span>.hifi.fasta</span><br></pre></td></tr></table></figure>

<p>上述三个内容都可以在同一文件夹下，随后在该文件夹下创建 <strong>fastq</strong> 文件夹，并将HIC数据放在 <strong>fastq</strong> 文件夹中，然后就可以运行juicer了。<br>！！！ <strong>fastq</strong> 文件中的文件必须以 <strong>fastq.gz</strong> 结尾，否则无法识别。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">bash <span class="variable">$&#123;path_to_juicer&#125;</span>/scripts/juicer.sh \</span><br><span class="line">-d ./</span><br><span class="line">-g <span class="variable">$&#123;sp&#125;</span> \</span><br><span class="line">-D <span class="variable">$&#123;path_to_juicer&#125;</span> \</span><br><span class="line">-z <span class="variable">$&#123;sp&#125;</span>.hifi.fasta \</span><br><span class="line">-y <span class="variable">$&#123;sp&#125;</span>_DpnII.txt \</span><br><span class="line">-p <span class="variable">$&#123;sp&#125;</span>.lens \</span><br><span class="line">-s DpnII \</span><br><span class="line">-t 20 \</span><br><span class="line">--assembly</span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">-d top-level路径，在该路径下应该包含fastq文件夹，且输出文件aligned和splits也会在该路径下</span><br><span class="line">-g 基因组的名字，这个可以随便取</span><br><span class="line">-D juicer软件路径，在该路径下要包含 scripts文件夹</span><br><span class="line">-z bwa索引的前缀</span><br><span class="line">-y 基因组酶切位点文件的路径</span><br><span class="line">-p 基因组长度文件的路径</span><br><span class="line">-s 用于做HIC的酶切酶名称</span><br><span class="line">-t threads，运行该命令的线程数</span><br><span class="line">--assembly 用以输出3D-DNA所需的 merged_nodups文件（juicer1.6不需要加这个）</span><br></pre></td></tr></table></figure>

<p>在 <strong>inter.txt</strong> 文件中可以看到HIC文件的比对率以及重复率等统计结果，用以评判数据质量是否符合要求，数据是否需要多测</p>
<h2 id="2-3-3D-DNA挂载"><a href="#2-3-3D-DNA挂载" class="headerlink" title="2.3 3D-DNA挂载"></a>2.3 3D-DNA挂载</h2><p>3D-DNA的输入文件比较简单，只需要输入参考基因组（hifi拼接出来的contig文件）以及juicer比对结果文件（merged_nodups.txt）即可</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">bash <span class="variable">$&#123;path_to_3D-DNA&#125;</span>/run-asm-pipeline.sh \</span><br><span class="line">-r 2 \</span><br><span class="line"><span class="variable">$&#123;sp&#125;</span>.hifi.fasta \</span><br><span class="line">merged_nodups.txt</span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">-r 校正连接的迭代次数</span><br></pre></td></tr></table></figure>

<p>最终3D-DNA输出的hic文件可以用juicerbox打开，并手动矫正，可以参考<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/BV1xD4y1m712/?vd_source=96abe912d50da25b76af5c20d7a0979e">juicerbox教程</a>,也可以参考<a target="_blank" rel="noopener" href="https://blog.csdn.net/u012110870/article/details/115511969">徐洲更老师的博客</a></p>
<h1 id="3-数据评估"><a href="#3-数据评估" class="headerlink" title="3. 数据评估"></a>3. 数据评估</h1><p>根据上述过程中的组装和比对结果，可以对拼接得到的基因组有一个大概的了解：染色体条数，基因组大小等。<br>在我个人组装的过程中存在以下问题：</p>
<ol>
<li>HIC比对率低(30%-40%)，PCR重复率高(60%-70%)，而在师姐之前的实验记录中，她的测序深度大约为我的10倍，比对率在70%左右，PCR重复&lt;20%。</li>
<li>3D-DNA挂载结果不佳，染色体间与染色体内的互作强度相近，软件几乎无法区分出染色体。</li>
</ol>
<p>为排除可能的因素，首先测试了比对软件，选用了 <strong>chromap+yahs</strong> 的方法去进行比对挂载，发现问题仍然存在。因此，有可能是数据文件的质量不佳，且初步怀疑是 <strong>HIC文库复杂度低</strong> 导致的。</p>
<p>考虑到目前已经获取了对应物种的转录组数据，因此计划使用 转录组与hifi组装结果进行比对 ，确保hifi数据不存在问题。</p>
<p>而常用的转录组比对软件STAR需要基因组注释文件，组装的contig基因组尚未进行基因组注释。故选用hisat2对转录组和hifiasm结果文件进行比对测试。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#建立hisat2比对索引</span></span><br><span class="line">$ hisat2-build -p 10 \  <span class="comment">#线程数</span></span><br><span class="line"><span class="variable">$&#123;sp&#125;</span>.hifi.fasta  \  <span class="comment">#hifiasm比对结果文件</span></span><br><span class="line"><span class="variable">$&#123;sp&#125;</span>_hisat_idx   <span class="comment">#输出索引的前缀</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#hisat2比对</span></span><br><span class="line">hisat2 -x <span class="variable">$&#123;sp&#125;</span>_hisat_idx \   <span class="comment">#基因组索引</span></span><br><span class="line">-1 <span class="variable">$&#123;path_to_hic&#125;</span>/<span class="variable">$&#123;sp&#125;</span>_HIC_R1.fq.gz \  <span class="comment">#HIC_reads1</span></span><br><span class="line">-2 <span class="variable">$&#123;path_to_hic&#125;</span>/<span class="variable">$&#123;sp&#125;</span>_HIC_R2.fq.gz  \   <span class="comment">#HIC_reads2</span></span><br><span class="line">-S <span class="variable">$&#123;sp&#125;</span>_hisat.sam \   <span class="comment">#输出结果文件</span></span><br><span class="line">-p 10   <span class="comment">#线程数</span></span><br></pre></td></tr></table></figure>

<p>hisat2的比对结果展示了 &gt;90% 的转录组reads 能够比对上hifiasm的组装结果文件，可以交叉验证这两组数据来自于同一物种且具有较高的质量，这证明了： <strong>HIFI数据质量可靠，HIC数据可能存在较大的质量问题</strong></p>

      </div>
      
      
      
    </div>
    

    
    
    


          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#Introduction"><span class="nav-number">1.</span> <span class="nav-text">Introduction</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#1-Install-software"><span class="nav-number">2.</span> <span class="nav-text">1. Install software</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#1-1-%E5%AE%89%E8%A3%85hifiasm"><span class="nav-number">2.1.</span> <span class="nav-text">1.1 安装hifiasm</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#1-1-1-conda%E5%AE%89%E8%A3%85"><span class="nav-number">2.1.1.</span> <span class="nav-text">1.1.1 conda安装</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#1-1-2-github%E5%AE%89%E8%A3%85"><span class="nav-number">2.1.2.</span> <span class="nav-text">1.1.2 github安装</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#1-2-%E5%AE%89%E8%A3%85AGB-%EF%BC%88%E5%B0%86gfa%E8%BD%AC%E4%B8%BAfa%E6%96%87%E4%BB%B6-%E5%8F%AF%E9%80%89%EF%BC%89"><span class="nav-number">2.2.</span> <span class="nav-text">1.2 安装AGB （将gfa转为fa文件,可选）</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#1-3-juicer%E5%AE%89%E8%A3%85"><span class="nav-number">2.3.</span> <span class="nav-text">1.3 juicer安装</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#1-4-3D-DNA%E5%AE%89%E8%A3%85"><span class="nav-number">2.4.</span> <span class="nav-text">1.4 3D-DNA安装</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#2-%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85"><span class="nav-number">3.</span> <span class="nav-text">2. 基因组组装</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#2-1-hifiasm%E6%8B%BC%E6%8E%A5"><span class="nav-number">3.1.</span> <span class="nav-text">2.1 hifiasm拼接</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-2-juicer%E6%AF%94%E5%AF%B9"><span class="nav-number">3.2.</span> <span class="nav-text">2.2 juicer比对</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-3-3D-DNA%E6%8C%82%E8%BD%BD"><span class="nav-number">3.3.</span> <span class="nav-text">2.3 3D-DNA挂载</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#3-%E6%95%B0%E6%8D%AE%E8%AF%84%E4%BC%B0"><span class="nav-number">4.</span> <span class="nav-text">3. 数据评估</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
  <p class="site-author-name" itemprop="name">John Doe</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">8</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
        <span class="site-state-item-count">3</span>
        <span class="site-state-item-name">categories</span>
      </div>
      <div class="site-state-item site-state-tags">
        <span class="site-state-item-count">19</span>
        <span class="site-state-item-name">tags</span>
      </div>
  </nav>
</div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2024</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">John Doe</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://muse.theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Muse</a>
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/muse.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

</body>
</html>
